### ISSUE TITLE
Discussion: Loops / Iterations / Ranges

### ISSUE THREAD

**Issue by ThePixelDeveloper** [newcat/baklavajs#162]
Morning 👋🏼 

I'm opening this to start a discussion on working with loops, iterations and ranges. I'm trying to work on something like [Node Red's Loops](https://flows.nodered.org/node/node-red-contrib-loop)

![CleanShot 2021-08-24 at 15 10 03@2x](https://user-images.githubusercontent.com/47728/130632025-949da40a-8cbe-43c5-add1-1750c8a1879c.png)

or the way Unreal Engine Blueprint does it: 

![CleanShot 2021-08-24 at 15 15 16@2x](https://user-images.githubusercontent.com/47728/130632951-7ab7dd0b-e801-4458-a32e-cb169f919089.png)

but struggling to come up with a solution for it within Baklava. 

--- 

The way I've currently implemented it, is to make each node to the right of the `Iterate` aware that it's dealing with an iterable value and treat it differently, but that feels like a hack. 

![CleanShot 2021-08-24 at 15 18 11@2x](https://user-images.githubusercontent.com/47728/130633361-4aa7de58-fd42-4cc8-88d1-7dfaf258a383.png)

I was wondering if any thought has been made into how to deal with this? I thinking I need to come up with a different way of executing the graph.

**Comment by newcat** [comment]
Yes indeed, the Baklava Engine was never designed to handle loops. I guess the best way to implement them is to create your own Engine. You can have a look at the [original Engine implementation](https://github.com/newcat/baklavajs/blob/master/packages/baklavajs-plugin-engine/src/engine.ts) and add the necessary parts for loops.

In Baklava v2 loops won't be supported out of the box either but they will be a lot easier to implement.

**Comment by ThePixelDeveloper** [comment]
Thanks for the reply @newcat; I appreciate the time and effort you put into it and it was the reason I sent you some monies. 💰 

I've been mulling it over and I'll be trying the approach of Unreal and Unity. They make a distinction which make sense to my monkey brain, and that's `execution flow` and `data` are separate concepts. It's easier to see in Unity here:


![](https://docs.unity3d.com/Packages/com.unity.visualscripting@1.7/manual/images/bolt-connections10.png)
[Source](https://docs.unity3d.com/Packages/com.unity.visualscripting@1.7/manual/vs-connections.html)

As you rightly pointed out, green arrows (execution) flow left to right and the data is gathered for the current node, from right to left. It's the reason I've been struggling to conceptualise it within Baklava JS.

So taking their [for each loop example](https://docs.unity3d.com/Packages/com.unity.visualscripting@1.7/manual/vs-control.html#for-each-loop):

![](https://docs.unity3d.com/Packages/com.unity.visualscripting@1.7/manual/images/bolt-controls7.png)

There's obviously some magic happening under the hood where the string nodes they're operating on the element of the collection, but that I'm not too worried about. 

--- 

> What would happen with the bottom trigger? It would execute the bottom Print String node, which is dependent on the String converter node which is in turn dependent on the ForLoop node. However, the ForLoop node has not been executed and never will be - so what value should "Index" have?

I don't have a copy of Unity / Unreal (maybe I should), but I have a gut feeling that's not possible. 

--- 

I'll let you know if I end up working on my own engine to support such flows.
